Some macros not generated with 7.2.x

Config:
```yaml
name: LibGLib
description: GLib
output: lib/bindings.dart
headers:
  entry-points:
    - 'gtype.h'
```

[C](https://github.com/GNOME/glib/blob/main/gobject/gtype.h):
```c
// gtype.h
#ifndef __G_TYPE_H__
#define __G_TYPE_H__

typedef unsigned long GType;

#define G_TYPE_INVALID G_TYPE_MAKE_FUNDAMENTAL(0)
#define G_TYPE_NONE G_TYPE_MAKE_FUNDAMENTAL(1)
#define G_TYPE_INTERFACE G_TYPE_MAKE_FUNDAMENTAL(2)
#define G_TYPE_CHAR G_TYPE_MAKE_FUNDAMENTAL(3)
// ...

#define G_TYPE_FUNDAMENTAL_SHIFT (2)
#define G_TYPE_MAKE_FUNDAMENTAL(x) ((GType)((x) << G_TYPE_FUNDAMENTAL_SHIFT))

#endif /* __G_TYPE_H__ */
```

Dart generated with ffigen 7.2.x:
```dart
// AUTO GENERATED FILE, DO NOT EDIT.
//
// Generated by `package:ffigen`.

const int G_TYPE_FUNDAMENTAL_SHIFT = 2;
```

vs. Dart generated with ffigen 7.1.0:
```dart
// AUTO GENERATED FILE, DO NOT EDIT.
//
// Generated by `package:ffigen`.

const int G_TYPE_INVALID = 0;

const int G_TYPE_NONE = 4;

const int G_TYPE_INTERFACE = 8;

const int G_TYPE_CHAR = 12;

const int G_TYPE_FUNDAMENTAL_SHIFT = 2;
```

Thanks for reporting this @jpnurmi 

We generate a temporary file for macros, but since #464 it's been generated at a different location than the working directory which has caused certain macros to not be generated properly.

I had to add a logging statement to get these (we don't log these because they are expected to contain errors from macros that we can't parse, but I guess we should at least generate a FINE log)
```red
[SEVERE] : Header /var/folders/x6/n78t6_tj49g06rbv445cpr0r0000gp/T/zpbWqV/temp_for_macros.hpp: Total errors/warnings: 7.
[SEVERE] :     /var/folders/x6/n78t6_tj49g06rbv445cpr0r0000gp/T/zpbWqV/temp_for_macros.hpp:1:10: error: './headers/example.h' file not found [Lexical or Preprocessor Issue]
[SEVERE] :     /var/folders/x6/n78t6_tj49g06rbv445cpr0r0000gp/T/zpbWqV/temp_for_macros.hpp:1:10: error: './headers/example.h' file not found [Lexical or Preprocessor Issue]
[SEVERE] :     /var/folders/x6/n78t6_tj49g06rbv445cpr0r0000gp/T/zpbWqV/temp_for_macros.hpp:2:62: error: expected expression [Parse Issue]
[SEVERE] :     /var/folders/x6/n78t6_tj49g06rbv445cpr0r0000gp/T/zpbWqV/temp_for_macros.hpp:3:52: error: use of undeclared identifier 'GType' [Semantic Issue]
[SEVERE] :     /var/folders/x6/n78t6_tj49g06rbv445cpr0r0000gp/T/zpbWqV/temp_for_macros.hpp:4:49: error: use of undeclared identifier 'GType' [Semantic Issue]
[SEVERE] :     /var/folders/x6/n78t6_tj49g06rbv445cpr0r0000gp/T/zpbWqV/temp_for_macros.hpp:5:54: error: use of undeclared identifier 'GType' [Semantic Issue]
[SEVERE] :     /var/folders/x6/n78t6_tj49g06rbv445cpr0r0000gp/T/zpbWqV/temp_for_macros.hpp:6:49: error: use of undeclared identifier 'GType' [Semantic Issue]
```

As a workaround I changed `TEST_TMPDIR` environment variable to the working directory, but macro generation still failed, with a semantic issue
```red
[SEVERE] :     /Users/prerakmann/development/ffigen/example/simple/temp_for_macros.hpp:6:49: error: use of undeclared identifier 'GType' [Semantic Issue]
```

Seems like supplying `-imacro/path/to/header` to clang marks the include-guards as used as well? since only if I remove them (`#ifndef __G_TYPE_H__`) I am able to parse macros

cc @TzviPM @dcharkes

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,3 +1,7 @@
+# 7.2.7
+
+- Fix some macros not being generated in some cases due to relative header paths.
+
 # 7.2.6
 
 - Fix path normalization behaviour for absolute paths and globs starting with `**`.

--- a/lib/src/header_parser/sub_parsers/macro_parser.dart
+++ b/lib/src/header_parser/sub_parsers/macro_parser.dart
@@ -68,9 +68,6 @@ List<Constant>? parseSavedMacros() {
   var cmdLen = 0;
 
   final compilerOpts = config.compilerOpts;
-  config.headers.entryPoints.followedBy([file.path]).forEach((entryFile) {
-    compilerOpts.add("-imacros$entryFile");
-  });
   clangCmdArgs = createDynamicStringArray(compilerOpts);
 
   cmdLen = config.compilerOpts.length;
@@ -205,7 +202,8 @@ File createFileForMacros() {
   // Write file contents.
   final sb = StringBuffer();
   for (final h in config.headers.entryPoints) {
-    sb.writeln('#include "$h"');
+    final fullHeaderPath = File(h).absolute.path;
+    sb.writeln('#include "$fullHeaderPath"');
   }
 
   _macroVarNames = {};

--- a/lib/src/header_parser/sub_parsers/macro_parser.dart
+++ b/lib/src/header_parser/sub_parsers/macro_parser.dart
@@ -84,6 +84,7 @@ List<Constant>? parseSavedMacros() {
   if (tu == nullptr) {
     _logger.severe('Unable to parse Macros.');
   } else {
+    logTuDiagnostics(tu, _logger, file.path, logLevel: Level.FINEST);
     final rootCursor = clang.clang_getTranslationUnitCursor(tu);
 
     final resultCode = clang.clang_visitChildren(

--- a/lib/src/header_parser/utils.dart
+++ b/lib/src/header_parser/utils.dart
@@ -30,16 +30,13 @@ void visitChildrenResultChecker(int resultCode) {
 
 /// Logs the warnings/errors returned by clang for a translation unit.
 void logTuDiagnostics(
-  Pointer<clang_types.CXTranslationUnitImpl> tu,
-  Logger logger,
-  String header,
-) {
+    Pointer<clang_types.CXTranslationUnitImpl> tu, Logger logger, String header,
+    {Level logLevel = Level.SEVERE}) {
   final total = clang.clang_getNumDiagnostics(tu);
   if (total == 0) {
     return;
   }
-
-  logger.severe('Header $header: Total errors/warnings: $total.');
+  logger.log(logLevel, 'Header $header: Total errors/warnings: $total.');
   for (var i = 0; i < total; i++) {
     final diag = clang.clang_getDiagnostic(tu, i);
     final cxstring = clang.clang_formatDiagnostic(
@@ -50,7 +47,7 @@ void logTuDiagnostics(
           clang_types
               .CXDiagnosticDisplayOptions.CXDiagnostic_DisplayCategoryName,
     );
-    logger.severe('    ' + cxstring.toStringAndDispose());
+    logger.log(logLevel, '    ' + cxstring.toStringAndDispose());
     clang.clang_disposeDiagnostic(diag);
   }
 }

--- a/pubspec.yaml
+++ b/pubspec.yaml
@@ -3,7 +3,7 @@
 # BSD-style license that can be found in the LICENSE file.
 
 name: ffigen
-version: 7.2.6
+version: 7.2.7
 description: Generator for FFI bindings, using LibClang to parse C header files.
 repository: https://github.com/dart-lang/ffigen
 

--- a/test/header_parser_tests/macros.h
+++ b/test/header_parser_tests/macros.h
@@ -30,3 +30,16 @@
 #define TEST14 INFINITY
 #define TEST15 -INFINITY
 #define TEST16 NAN
+
+
+// gtype.h
+#ifndef __TEST_INCLUDE_GUARD_1__
+#define __TEST_INCLUDE_GUARD_1__
+
+#define TEST17 TEST_FUNCTION_MACRO(0)
+#define TEST18 TEST_FUNCTION_MACRO(1)
+#define TEST19 TEST_FUNCTION_MACRO(2)
+
+#define TEST_FUNCTION_MACRO(x) ((unsigned long)((x) << (2)))
+
+#endif /* __TEST_INCLUDE_GUARD_1__ */

--- a/test/header_parser_tests/macros.h
+++ b/test/header_parser_tests/macros.h
@@ -32,7 +32,7 @@
 #define TEST16 NAN
 
 
-// gtype.h
+// Inside include guard.
 #ifndef __TEST_INCLUDE_GUARD_1__
 #define __TEST_INCLUDE_GUARD_1__
 

--- a/test/header_parser_tests/macros_test.dart
+++ b/test/header_parser_tests/macros_test.dart
@@ -98,6 +98,18 @@ ${strings.headers}:
       expect(actual.getBindingAsString('TEST16'),
           expected.getBindingAsString('TEST16'));
     });
+    test('TEST17', () {
+      expect(actual.getBindingAsString('TEST17'),
+          expected.getBindingAsString('TEST17'));
+    });
+    test('TEST18', () {
+      expect(actual.getBindingAsString('TEST18'),
+          expected.getBindingAsString('TEST18'));
+    });
+    test('TEST19', () {
+      expect(actual.getBindingAsString('TEST19'),
+          expected.getBindingAsString('TEST19'));
+    });
   });
 }
 
@@ -125,6 +137,9 @@ Library expectedLibrary() {
           rawType: 'double',
           rawValue: strings.doubleNegativeInfinity),
       Constant(name: 'TEST16', rawType: 'double', rawValue: strings.doubleNaN),
+      Constant(name: 'TEST17', rawType: 'int', rawValue: "0"),
+      Constant(name: 'TEST18', rawType: 'int', rawValue: "4"),
+      Constant(name: 'TEST19', rawType: 'int', rawValue: "8"),
     ],
   );
 }

